feat(Idempotency): add feature for manipulating idempotent responses

walmsles opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number:** #2164

## Summary

I have added the capability for developers to provide a response hook, which is called when the idempotency handler finds a valid idempotent response.

### Changes

Addition of "response_hook" in IdempotentConfig class to enable developers to register a repsonse_hook

### User experience

```python
import datetime
import uuid
from typing import Dict

from aws_lambda_powertools import Logger
from aws_lambda_powertools.utilities.idempotency import (
    DynamoDBPersistenceLayer,
    IdempotencyConfig,
    idempotent_function,
)
from aws_lambda_powertools.utilities.idempotency.persistence.base import (
    DataRecord,
)
from aws_lambda_powertools.utilities.typing import LambdaContext

logger = Logger()

def my_response_hook(response: Dict, idempotent_data: DataRecord) -> Dict:
    # Return inserted Header data into the Idempotent Response
    response["x-idempotent-key"] = idempotent_data.idempotency_key

    # expiry_timestamp could be None so include if set
    expiry_timestamp = idempotent_data.expiry_timestamp
    if expiry_timestamp:
        expiry_time = datetime.datetime.fromtimestamp(int(expiry_timestamp))
        response["x-idempotent-expiration"] = expiry_time.isoformat()

    # Must return the response here
    return response

dynamodb = DynamoDBPersistenceLayer(table_name="IdempotencyTable")
config = IdempotencyConfig(response_hook=my_response_hook)

@idempotent_function(data_keyword_argument="order", config=config, persistence_store=dynamodb)
def process_order(order: dict) -> dict:
    # create the order_id
    order_id = str(uuid.uuid4())

    # create your logic to save the order
    # append the order_id created
    order["order_id"] = order_id

    # return the order
    return {"order": order}


def lambda_handler(event: dict, context: LambdaContext):
    config.register_lambda_context(context)  # see Lambda timeouts section
    try:
        logger.info(f"Processing order id {event.get('order_id')}")
        return process_order(order=event.get("order"))
    except Exception as err:
        return {"status_code": 400, "error": f"Erro processing {str(err)}"}
```
### Developer experience and implementation


#### Response hooks
We are introducing support for hooks in the Idempotency utility. Customers can now create a function and pass it as the hook in the IdempotencyConfig. This hook will be invoked every time a request is idempotent. If the request is not idempotent, the hook will not execute.

```python
from aws_lambda_powertools.utilities.idempotency import IdempotencyConfig
def my_response_hook(response: Dict, idempotent_data: DataRecord): ...
config = IdempotencyConfig(response_hook=my_response_hook)
```

#### Manipulating response
This hook will be able to access both: the response from the invocation and the `DataRecord` from the Idempotency Store. This means customers can inject new keys into the response or utilize attributes from the DataRecord to log information or create metrics, for example.

#### Persistent Store
This hook is compatible with any Persistent Store. Currently, we support DynamoDB and Redis for this implementation.

#### Handling exceptions
Customers are responsible for handling exceptions when using Response hooks. During the implementation of this new feature, we considered the possibility of implementing soft failure or raising warnings to ignore hook execution. For instance, this approach could prevent customers from having exceptions in the hook, thereby avoiding potential malfunctions in the Lambda function execution. However, we opted against this approach because it could result in unexpected behaviors and potentially breaking the contract that customers expect for `Response hooks`. Let me provide some context with an example.
Consider a scenario where a customer creates a hook to inject a specific key into the response when the operation is idempotent and wants to use it to trigger an alert for third-party integration. The code might appear like this:

```python
def my_response_hook(response: Dict, idempotent_data: DataRecord) -> Dict:
    # Return inserted Header data into the Idempotent Response
    response["x-idempotent-key"] = idempotent_data.idempotency_key

    imagine_an_exception_here()

    # expiry_timestamp could be None so include if set
    expiry_timestamp = idempotent_data.expiry_timestamp
    if expiry_timestamp:
        expiry_time = datetime.datetime.fromtimestamp(int(expiry_timestamp))
        response["x-idempotent-expiration"] = expiry_time.isoformat()

    # Must return the response here
    return response

dynamodb = DynamoDBPersistenceLayer(table_name="IdempotencyTable")
config = IdempotencyConfig(response_hook=my_response_hook)

@idempotent_function(data_keyword_argument="order", config=config, persistence_store=dynamodb)
def process_order(order: dict) -> dict: ...

def lambda_handler(event: dict, context: LambdaContext):
    config.register_lambda_context(context)  # see Lambda timeouts section
    process_order = process_order(event.get("order"))

    if process_order["x-idempotent-key"] is not None:
        trigger_some_sns_alert()
```

If we choose to ignore the hook execution exception and only raise a warning instead of propagating the exception, customers may encounter difficulties in debugging. They may struggle to understand why the `if process_order["x-idempotent-key"] is not None:` condition is failing and why the `x-idempotent-key` doesn't exist in the response.

#### Documentation
The documentation will be updated to inform customers on the best practices for using Response hooks. Additionally, it will feature an example demonstrating how to utilize Response hooks and effectively handle exceptions.



## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary> No, not a breaking change.  Configuration is optional and implementation checked for existence of the hook


**RFC issue number**: 2164

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

walmsles committed:
feat(idempotent-response-manipulation): Added capability of providing an IdempotentHook functiont to be called when an idempotent response is being returned.

walmsles commented:
Still need to work on documentation - will do this next.

walmsles committed:
Merge branch 'develop' into feat-manip-idempotent-response

walmsles committed:
chore(mypy): resolve myopy static typing issues, make response+hook properly optional

leandrodamascena commented:
Hi @walmsles! Your contributions continue to shine brightly. Access to the Idempotency operation results has long been on our backlog and I'm so happy that we will add this new feature. If you need any assistance with documentation, testing, or anything else, please don't hesitate to let me know.

Thanks

walmsles committed:
feat(response_hook): added some documentation, call response_hook after custom de-serialization

walmsles commented:
I added documentation for review. I also changed the order so that response_hook processes **after** the custom de-serializer if it is used. This makes the most sense to developers so the response_hook is passed an expected structure rather than a raw event format.

leandrodamascena commented on aws_lambda_powertools/utilities/idempotency/__init__.py:
```diff
@@ -2,6 +2,7 @@
 Utility for adding idempotency to lambda functions
 """
 
+from aws_lambda_powertools.utilities.idempotency.config import IdempotentHookData, IdempotentHookFunction
```
Would it be better to move these classes to `hook.py` or something like that?

leandrodamascena commented on aws_lambda_powertools/utilities/idempotency/config.py:
```diff
@@ -1,8 +1,45 @@
 from aws_lambda_powertools.utilities.typing import LambdaContext
 
 
+class IdempotentHookData:
+    """
+    Idempotent Hook Data
+
+    Contains data relevant to the current Idempotent record which matches the current request.
+    All IdempotentHook functions will be passed this data as well as the current Response.
+    """
+
+    def __init__(self, data_record: DataRecord) -> None:
+        self._idempotency_key = data_record.idempotency_key
+        self._status = data_record.status
+        self._expiry_timestamp = data_record.expiry_timestamp
+
+    @property
+    def idempotency_key(self) -> str:
+        return self._idempotency_key
+
+    @property
+    def status(self) -> str:
+        return self._status
+
+    @property
+    def expiry_timestamp(self) -> Optional[int]:
+        return self._expiry_timestamp
```
I'm considering adding a field named `raw_data_record` to return the `data_record` as JSON. We can't predict if the customer might access  `in_progress_expiry_timestamp` or `payload_hash`, for instance. Make sense?

rubenfonseca replied:
I kinda agree, but if we expose `raw_data_record` why do we need this layer of indirection in the first place? Can't we just send the entire DataRecord directly on the hook functions?

leandrodamascena commented on aws_lambda_powertools/utilities/idempotency/__init__.py:
```diff
@@ -2,6 +2,7 @@
 Utility for adding idempotency to lambda functions
 """
 
+from aws_lambda_powertools.utilities.idempotency.config import IdempotentHookData, IdempotentHookFunction
```
Would it be better to move these classes to `hook.py` or something like that?

walmsles replied:
Maybe let's jump on a call - placed where they are to avoid circular references between base.py, config.py and hook.py - happy for guidance  to improve - that's where they started.

leandrodamascena replied:
Sure thing! Let's schedule a call to review and potentially merge this PR before the next release. I'll ping you on Discord and we can find a slot to join in a meeting.

Just out of curiosity, why would we have issues with circular references here? We're importing from different files, so it shouldn't be a problem. I might be missing something, but I can't see any potential problems with circular references here.

I've created a patch with this modification, and all tests are passing, mypy is also happy, and I couldn't spot any issues with circular references. 
Could you please try applying this patch in your local environment?

[hook_patch.txt](https://github.com/aws-powertools/powertools-lambda-python/files/14844033/hook_patch.txt)

leandrodamascena commented on docs/utilities/idempotency.md:
```diff
@@ -909,6 +910,22 @@ You can create your own persistent store from scratch by inheriting the `BasePer
 
     For example, the `_put_record` method needs to raise an exception if a non-expired record already exists in the data store with a matching key.
 
+### Modifying the Idempotent Repsonse
```
Not everyone will want to modify the response. Perhaps `Accessing the Idempotent Response` would be more appropriate?

rubenfonseca replied:
What about `Manipulating the Idempotent Response`?

leandrodamascena reviewed (changes requested):
Hey, @walmsles! I looked at the code and ran some tests, and everything looked good to me at first glance! I left some comments for us to discuss before a final review. Thank you so much!

rubenfonseca commented on aws_lambda_powertools/utilities/idempotency/base.py:
```diff
@@ -227,6 +230,11 @@ def _handle_for_status(self, data_record: DataRecord) -> Optional[Any]:
             )
         response_dict: Optional[dict] = data_record.response_json_as_dict()
         if response_dict is not None:
+            if self.config.response_hook is not None:
+                return self.config.response_hook(
+                    self.output_serializer.from_dict(response_dict),
+                    IdempotentHookData(data_record),
+                )
             return self.output_serializer.from_dict(response_dict)
```
Suggestion just to keep the code cleaner

```suggestion
            serialized_output = self.output_serializer.from_dict(response_dict)
            if self.config.response_hook is not None:
                return self.config.response_hook(
                    serialized_output
                    IdempotentHookData(data_record),
                )
            return serialized_output
```

rubenfonseca commented on docs/utilities/idempotency.md:
```diff
@@ -699,15 +699,16 @@ For advanced configurations, such as setting up SSL certificates or customizing
+| **use_local_cache**             | `False` | Whether to locally cache idempotency results                                                                                                                                                                                        |
+| **local_cache_max_items**       | 256     | Max number of items to store in local cache                                                                                                                                                                                         |
+| **hash_function**               | `md5`   | Function to use for calculating hashes, as provided by [hashlib](https://docs.python.org/3/library/hashlib.html){target="_blank" rel="nofollow"} in the standard library.                                                           |
+| **response_hook**               | `None`  | Function to use for processing the stored Idempotent response.  This function hook is called when an already returned response is found.  See [Modifying The Idempotent Response](idempotency.md#modifying-the-idempotent-repsonse) |
```
```suggestion
| **response_hook**               | `None`  | Function to use for processing the stored Idempotent response.  This function hook is called when an existing idempontent response is returned.  See [Modifying The Idempotent Response](idempotency.md#modifying-the-idempotent-repsonse) |
```

rubenfonseca commented on examples/idempotency/src/working_with_response_hook.py:
```diff
@@ -0,0 +1,27 @@
+from aws_lambda_powertools.utilities.typing import LambdaContext
+
+
+def my_response_hook(response: Dict, idempotent_data: IdempotentHookData) -> Dict:
+    # How to add a field to the response
+    response["is_idempotent_response"] = True
+
+    # Must return the response here
+    return response
```
Can we expand the sample to actually make use of `idempotent_data`?

walmsles committed:
feat(response_hook): review items

walmsles committed:
Merge branch 'develop' into feat-manip-idempotent-response

walmsles committed:
chore(mypy): resolve type erro r in example code - expiry_timestamp can be None

walmsles committed:
chore(docs): fix formatting error in markdown

walmsles committed:
chore(docs): fix highlighting of example code - lines moved

leandrodamascena committed:
Improving doc

leandrodamascena committed:
Improving doc

leandrodamascena commented:
Hey @walmsles! I've updated the documentation and example. I believe it's ready for the final review now.

I'll also be updating the issue with the decisions regarding exception/error handling, including why we allow customers to handle their own exceptions and why we fail in case of an exception.

walmsles commented:
I love the **being a Good Citizen** sections!  That makes so much more sense - I knew it was missing something.

rubenfonseca commented:
Reviewing now

rubenfonseca commented on examples/idempotency/src/working_with_response_hook.py:
```diff
@@ -1,27 +1,56 @@
+        logger.info(f"Processing order id {event.get('order_id')}")
+        return process_order(order=event.get("order"))
+    except Exception as err:
+        return {"status_code": 400, "error": f"Erro processing {str(err)}"}
```
```suggestion
        return {"status_code": 400, "error": f"Error processing {str(err)}"}
```

I loved this example, you can really understand how powerful the feature can be

rubenfonseca commented on docs/utilities/idempotency.md:
```diff
@@ -454,6 +454,40 @@ sequenceDiagram
+        deactivate Persistence Layer
+        Note over Response hook,Persistence Layer: Record status is COMPLETE and not expired
+        Response hook->>Lambda: Response hook invoked
+        Lambda-->>Client: Same response sent to client
```
```suggestion
        Lambda-->>Client: Manipulated idempotent response sent to client
```

Otherwise it makes me thing that the Response Hook can't change anything

rubenfonseca commented on docs/utilities/idempotency.md:
````diff
@@ -454,6 +454,40 @@ sequenceDiagram
+        Lambda-->>Client: Same response sent to client
+    end
+```
+<i>Idempotent successful request with response hook</i>
````
```suggestion
<i>Successful idempotent request with a response hook</i>
```

rubenfonseca commented on docs/utilities/idempotency.md:
```diff
@@ -910,22 +944,38 @@ You can create your own persistent store from scratch by inheriting the `BasePer
+### Manipulating the Idempotent Response
 
-You can provide the response_hook using _**IdempotentConfig**_.
+You can set up a `response_hook` in the `IdempotentConfig` class to access the returned data when an operation is idempotent. The hook function will be called with the current deserialized response object and the Idempotency record.
```
```suggestion
You can set up a `response_hook` in the `IdempotentConfig` class to manipulate the returned data when an operation is idempotent. The hook function will be called with the current deserialized response object and the Idempotency record.
```

rubenfonseca commented on docs/utilities/idempotency.md:
```diff
@@ -910,22 +944,38 @@ You can create your own persistent store from scratch by inheriting the `BasePer
 
+#### Being a good citizen
+
+Using Response hooks can add subtle improvements to manipulating returned data from idempotent operations, but also add significant complexity if you're not careful.
+
+Keep the following in mind when authoring hooks for Idempotency utility:
+
+1. **Response hook works exclusively when operations are idempotent.** Carefully consider the logic within the `Response hook` and prevent any attempt to access the key from relying exclusively on idempotent operations.
+
+2. **Catch your own exceptions.** Catch and handle known exceptions to your logic.
+
+3. **Watch out when you are decorating the Lambda Handler and using the Response hook.** If you don't catch and handle exceptions in your `Response hook`, your function might not run properly.
```
A suggestion to make this more clear:

1- response hooks only work on idempotent responses
2- catch exceptions
3- keep it simple

Is this better?

```suggestion
When using response hooks to manipulate returned data from idempotent operations, it's important to follow best practices to avoid introducing complexity or issues. Keep these guidelines in mind:

1. **Response hook works exclusively when operations are idempotent.** The hook will not be called when an operation is not idempotent, or when the idempotent logic fails.

2. **Catch and Handle Exceptions** Your response hook code should catch and handle any exceptions that may arise from your logic. Unhandled exceptions will cause the Lambda function to fail unexpectedly.

3. **Keep Hook Logic Simple** Response hooks should contain minimal, straightforward logic for manipulating response data. Avoid complex conditional branching and aim for hooks that are easy to reason about.
```

leandrodamascena replied:
This is much better than the one I wrote! Thank you for this very useful review and suggestion @rubenfonseca

leandrodamascena committed:
Addressing Ruben's feedback

leandrodamascena commented:
@rubenfonseca ready to one more review.

rubenfonseca commented:
Great job @walmsles @leandrodamascena pulling this off!
